syntax = "proto3";

package cust_context;

// Customer Context Service - Optimized Version (7 Working Methods Only)
service CustContextService {
    // ========================================
    // TIER 1: CONTEXT MANAGEMENT (2 methods)
    // ========================================
    rpc CreateContext(CreateContextRequest) returns (CreateContextResponse);
    rpc UpdateContext(UpdateContextRequest) returns (UpdateContextResponse);
    
    // ========================================
    // TIER 2: EMOTIONAL & INTENT INTELLIGENCE (2 methods)
    // ========================================
    rpc SetConversationMood(SetConversationMoodRequest) returns (SetConversationMoodResponse);
    rpc TrackUserIntent(TrackUserIntentRequest) returns (TrackUserIntentResponse);
    
    // ========================================
    // TIER 3: CONVERSATION ANALYSIS (3 methods)
    // ========================================
    rpc GetConversationFlow(GetConversationFlowRequest) returns (GetConversationFlowResponse);
    rpc PredictNextUserQuestion(PredictNextUserQuestionRequest) returns (PredictNextUserQuestionResponse);
    rpc DisambiguateEntity(DisambiguateEntityRequest) returns (DisambiguateEntityResponse);
}

// ========================================
// CONTEXT MANAGEMENT MESSAGES
// ========================================

// Create new conversation context
message CreateContextRequest {
    string session_id = 1;
    string tenant_id = 2;
    int32 ttl_seconds = 3;
}

message CreateContextResponse {
    bool success = 1;
    string message = 2;
    string session_id = 3;
    int64 created_timestamp = 4;
}

// Update existing conversation context - COMPLETE FIELD SET
message UpdateContextRequest {
    string session_id = 1;
    string tenant_id = 2;        // ADDED: API Gateway uses tenant_id
    string user_query = 3;       // VERIFIED: API Gateway uses user_query
    string entities = 4;         // VERIFIED: API Gateway uses entities as string
}

message UpdateContextResponse {
    bool success = 1;
    string message = 2;
}

// Supporting data structures (kept for compatibility)
message ConversationContext {
    string session_id = 1;
    string user_id = 2;
    string tenant_id = 3;
    repeated ConversationEntity entities = 4;
    repeated ConversationTurn turns = 5;
    int64 created_at = 6;
    int64 updated_at = 7;
}

message ConversationEntity {
    string entity_type = 1;
    string entity_value = 2;
    string entity_text = 3;
    double confidence = 4;
}

message ConversationTurn {
    string user_message = 1;
    string bot_response = 2;
    int64 timestamp = 3;
    repeated ConversationEntity extracted_entities = 4;
}

// ========================================
// EMOTIONAL & INTENT INTELLIGENCE MESSAGES
// ========================================

message SetConversationMoodRequest {
    string session_id = 1;
    string tenant_id = 2;
    string mood = 3;  // happy, confused, frustrated, excited, neutral
    string reason = 4;
    double confidence = 5;
}

message SetConversationMoodResponse {
    bool success = 1;
    string message = 2;
    string previous_mood = 3;
    string detected_mood = 4;
    double mood_confidence = 5;
}

message TrackUserIntentRequest {
    string session_id = 1;
    string tenant_id = 2;
    string intent = 3;  // browsing, buying, asking, complaining, comparing
    double confidence = 4;
    string detected_from = 5;
}

message TrackUserIntentResponse {
    bool success = 1;
    string message = 2;
    string recommended_response_style = 3;
    repeated string intent_history = 4;
}

// ========================================
// CONVERSATION ANALYSIS MESSAGES
// ========================================

message GetConversationFlowRequest {
    string session_id = 1;
    string tenant_id = 2;
}

message GetConversationFlowResponse {
    bool success = 1;
    repeated string topic_progression = 2;
    string current_topic = 3;
    string suggested_next_topic = 4;
    double flow_quality_score = 5;
}

message PredictNextUserQuestionRequest {
    string session_id = 1;
    string tenant_id = 2;
}

message PredictNextUserQuestionResponse {
    bool success = 1;
    repeated string likely_questions = 2;
    repeated string suggested_proactive_responses = 3;
    double prediction_confidence = 4;
}

message DisambiguateEntityRequest {
    string session_id = 1;
    string tenant_id = 2;
    string ambiguous_entity = 3;
    string context_hint = 4;
}

message DisambiguateEntityResponse {
    bool success = 1;
    string clarified_entity = 2;
    string entity_type = 3;
    double disambiguation_confidence = 4;
    repeated string alternative_interpretations = 5;
}